sdc/datatypes/hpat_pandas_dataframe_functions.py

```diff
@@ -31,64 +31,340 @@
+
+
+    @overload_method(DataFrameType, 'count')
+    def sdc_pandas_dataframe_count(self, axis=0, level=None, numeric_only=False):
+        """
+        Pandas DataFrame method :meth:`pandas.DataFrame.count` implementation.
+
+        .. only:: developer
+
+            Test: python -m sdc.runtests sdc.tests.test_dataframe.TestDataFrame.test_count
+
+        Parameters
+        -----------
+        self: :class:`pandas.DataFrame`
+            input arg
+        axis:
+            *unsupported*
+        level:
+            *unsupported*
+        numeric_only:
+            *unsupported*
+
+        Returns
+        -------
+        :obj:`pandas.Series` or `pandas.DataFrame`
+                returns: For each column/row the number of non-NA/null entries. If level is specified returns a DataFrame.
+        """
+
+        _func_name = 'Method pandas.dataframe.count().'
+
+        if not isinstance(self, DataFrameType):
+            raise TypingError('{} The object must be a pandas.dataframe. Given: {}'.format(_func_name, self))
+
+        if not (isinstance(axis, types.Omitted) or axis == 0):
+            raise TypingError("{} 'axis' unsupported. Given: {}".format(_func_name, axis))
+
+        if not (isinstance(level, types.Omitted) or level is None):
+            raise TypingError("{} 'level' unsupported. Given: {}".format(_func_name, axis))
+
+        if not (isinstance(numeric_only, types.Omitted) or numeric_only is False):
+            raise TypingError("{} 'numeric_only' unsupported. Given: {}".format(_func_name, axis))
+
+        def sdc_pandas_dataframe_count_impl(self, axis=0, level=None, numeric_only=False):
+            result_data = []
+            result_index = []
+
+            for dataframe_item in self._data:
+                item_count = dataframe_item.count()
+                item_name = dataframe_item._name
+                result_data.append(item_count)
+                result_index.append(item_name)
+
+            return pandas.Series(data=result_data, index=result_index)
+
+        return sdc_pandas_dataframe_count_impl
+
+else:
+    def reduce(df, name):
+        saved_columns = df.columns
+        n_cols = len(saved_columns)
+        data_args = tuple('data{}'.format(i) for i in range(n_cols))
+        func_text = "def _reduce_impl(df, axis=None, skipna=None, level=None, numeric_only=None):\n"
+        for i, d in enumerate(data_args):
+            func_text += "  {} = hpat.hiframes.api.init_series(hpat.hiframes.pd_dataframe_ext.get_dataframe_data(df, {}))\n".format(
+                d + '_S', i)
+            func_text += "  {} = {}.{}()\n".format(d + '_O', d + '_S', name)
+        func_text += "  data = np.array(({},))\n".format(
+            ", ".join(d + '_O' for d in data_args))
+        func_text += "  index = hpat.str_arr_ext.StringArray(({},))\n".format(
+            ", ".join("'{}'".format(c) for c in saved_columns))
+        func_text += "  return hpat.hiframes.api.init_series(data, index)\n"
```

Comment:
You concatenate strings via operator `+`, but more effective way is to create `list`, fill the list via `.append()` (do NOT forget to cut off `\n`), then join the list to the string via `'\n'.join()`, e.g.:
```
func_lines = []
func_lines.append('the first line')
func_lines.append('the second line')
func_lines.append('the last line')
func_text = '\n'.join(func_lines)
```